lua/neotest-haskell/base.lua

```diff
@@ -21,6 +22,38 @@ end
 base.match_package_root_pattern = lib.files.match_root_pattern('*.cabal', 'package.yaml')
 
 ---@type fun(file:string):(string|nil)
-base.match_project_root_pattern = lib.files.match_root_pattern('cabal.project', 'stack.yaml')
+base.match_project_root_pattern = lib.files.match_root_pattern('cabal.project')
+
+local function directory_contains_file_matching(directory, patterns)
+  for _, pattern in ipairs(patterns) do
+    for _, file in ipairs(vim.fn.glob(Path:new(directory, pattern).filename, true, true)) do
+      if Path:new(file):exists() then
+        return true
+      end
+    end
+  end
+  return false
+end
+
+--- Select a build tool from the given list of build tools preferring the first that can be used.
+---@param project_root string root directory of project
+---@param build_tools table list of build tools; supported values: 'stack', 'cabal'
+---@return string first build_tool that can be used for the project
+base.select_build_tool = function(project_root, build_tools)
+  for _, build_tool in pairs(build_tools) do
+    if build_tool == 'stack' and directory_contains_file_matching(project_root, { 'stack.yaml' }) then
+      return 'stack'
+    elseif build_tool == 'cabal' and directory_contains_file_matching(project_root, { '*.cabal', 'cabal.project' }) then
+      return 'cabal'
+    end
+  end
+  error('Cannot run tests for configured build tools: ' .. vim.inspect(build_tools))
+end
+
+---@param project_root string root directory of project
+---@return boolean true if directory contains a file name `cabal.project`
+base.is_multi_package_project = function(project_root)
+  return directory_contains_file_matching(project_root, { 'cabal.project' })
+end
```

Comment:
I would put this in the `cabal` module.

For a general implementation, we would have to handle Stack projects, too (yagni).

Reply:
stack already works with this. I've checked out the hspec source code, added a stack yaml and ran some random test with it.

Reply:
Stack does not need a `cabal.project` file to be present. You can have Stack projects with only a `stack.yaml` in the project root and `package.yaml` files in the package roots.

`stack.build_command` differentiates between multi and single package projects by checking if the package and project root are the same.

Come to think of it, it's also possible for a `cabal.project` file to be present in a single-package project. So we should use the same logic in `cabal.build_command`.

Sorry for not realising this earlier.

Reply:
I've added test fixtures for that and I've tested them manually. Those tests should probably be automated, but I'm not terribly motivated to do that right now.

Reply:
The logic for checking if it's a multi package project is now duplicated in the `stack` module.

🤔 I'm starting to think the `cabal` and `stack` modules have become redundant, and we should just move all the logic for creating the command here.

But since this is working, let's get this merged and I'll take care of refactoring it later.